iroha/src/genesis.rs

```diff
@@ -41,29 +62,196 @@ impl GenesisTransaction {
-        let genesis_key_pair = KeyPair {
-            public_key: init_config.genesis_account_public_key.clone(),
-            private_key: init_config
-                .genesis_account_private_key
-                .clone()
-                .ok_or("genesis account private key is empty")?,
+        genesis_config: &GenesisConfiguration,
+    ) -> Result<Option<GenesisBlock>, String> {
+        if let Some(genesis_block_path) = &genesis_config.genesis_block_path {
+            let file = File::open(Path::new(genesis_block_path))
+                .map_err(|e| format!("Failed to open a genesis block file: {}", e))?;
+            let reader = BufReader::new(file);
+            let raw_block: RawGenesisBlock = serde_json::from_reader(reader)
+                .map_err(|e| format!("Failed to deserialize json from reader: {}", e))?;
+            let genesis_key_pair = KeyPair {
+                public_key: genesis_config.genesis_account_public_key.clone(),
+                private_key: genesis_config
+                    .genesis_account_private_key
+                    .clone()
+                    .ok_or("genesis account private key is empty")?,
+            };
+            Ok(Some(GenesisBlock {
+                transactions: raw_block
+                    .transactions
+                    .iter()
+                    .map(|raw_transaction| raw_transaction.sign_and_accept(&genesis_key_pair))
+                    .filter_map(Result::ok)
+                    .collect(),
+                wait_for_peers_retry_count: genesis_config.wait_for_peers_retry_count,
+                wait_for_peers_retry_period_ms: genesis_config.wait_for_peers_retry_period_ms,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Submits genesis transactions.
+    pub async fn submit(&self, sumeragi: Arc<RwLock<Sumeragi>>) -> Result<(), String> {
+        let genesis_topology = {
+            let sumeragi = sumeragi.read().await;
+            self.wait_for_peers(sumeragi.peer_id.clone(), sumeragi.network_topology.clone())
+                .await?
         };
-        Ok(GenesisBlock {
-            transactions: raw_block
-                .transactions
-                .iter()
-                .map(|raw_transaction| raw_transaction.sign_and_accept(&genesis_key_pair))
-                .filter_map(Result::ok)
-                .collect(),
-        })
+        log::info!("Initializing iroha using the genesis block.");
+        sumeragi
+            .write()
+            .await
+            .start_genesis_round(self.transactions.clone(), genesis_topology)
+            .await
+    }
+
+    /// Waits for a minimum number of `peers` needed for consensus to be online.
+    /// Returns [`InitializedNetworkTopology`] with the set A consisting of online peers.
+    async fn wait_for_peers(
+        &self,
+        this_peer_id: PeerId,
+        network_topology: InitializedNetworkTopology,
+    ) -> Result<InitializedNetworkTopology, String> {
```

Comment:
Looks more network\sumeragi related. Maybe `NetworkBlock` should be placed there.